### PR TITLE
[7.x] Add ability to share auto-wired services where matches a prefix

### PR DESCRIPTION
**Pull request by alquerci**
see https://github.com/laravel/ideas/issues/2248

### CODE REVIEW

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -195,9 +209,64 @@ public function resolved($abstract)
...
+
+        $this->isSharedPrefixesCompiled = true;
+    }
+
+    /**
+     * Gets compiled shared prefixes.
+     *
+     * @return array
+     */
+    public function getCompiledSharedPrefixes(): array
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php:265]: Remove the return type

File: src/Illuminate/Container/Container.php [hunk 2]
```diff
@@ -1325,4 +1394,20 @@ public function __set($key, $value)
     {
         $this[$key] = $value;
     }
+
+    /**
+     * Compiles shared prefixes to sort from longer to shorter.
+     */
+    private function compileSharedPrefixes(): void
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php:1401]: same

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Container/Container.php (hunk 1 above)
Comment (alquerci) [reply src/Illuminate/Container/Container.php]: @GrahamCampbell For what reason return type are not allowed?

I want to understand why.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

File: src/Illuminate/Container/Container.php (hunk 1 above)
Comment (gravataLonga) [reply src/Illuminate/Container/Container.php]: i'm curious about this...

**Pull request closed without merging**